File: frameworks/java/javalin/UnitTests/src/test/java/unittests/AuthorizedTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import io.javalin.testtools.JavalinTest;
import io.javalin.testtools.Response;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

class AuthorizedTests extends JavalinApp {

    private static final String TOKEN = Expected.settings().get("token").asText();

    private static final String WRONG_TOKEN = Expected.settings().get("wrongToken").asText();

    // rb:test authorized.allowed
    @Test
    @Tag("authorized.allowed")
    void theSettingsTokenIsLetThrough() {
        JavalinTest.test(app(), (server, client) -> {
            Response response = client.get("/authorized/small", request -> request.header("authorization", "Bearer " + TOKEN));

            assertEquals(200, response.code());
            Answer.is(Expected.json("items.small.json"), response);
        });
    }

    // rb:test authorized.denied
    @Test
    @Tag("authorized.denied")
    void aTokenOneCharacterOffIsForbidden() {
        JavalinTest.test(app(), (server, client) -> {
            Response response = client.get("/authorized/small", request -> request.header("authorization", "Bearer " + WRONG_TOKEN));

            assertEquals(403, response.code());
            assertEquals("Forbidden", response.body().string());
        });
    }

    @Test
    void noTokenIsForbiddenToo() {
        JavalinTest.test(app(), (server, client) -> assertEquals(403, client.get("/authorized/small").code()));
    }

    @Test
    void theCheckRunsOnAuthorizedAlone() {
        JavalinTest.test(app(), (server, client) ->
            assertEquals(200, client.get("/json/small", request -> request.header("authorization", "Bearer " + WRONG_TOKEN)).code()));
    }
}

File: frameworks/dotnet/hardened/UnitTests/AuthorizedTests.cs
namespace UnitTests;

public class AuthorizedTests
{
    private static string Token(string name) => (string)Expected.Settings[name]!;

    // rb:test authorized.allowed
    [ModuleTest]
    [Trait("corpus", "authorized.allowed")]
    public async Task The_settings_token_is_let_through(ITestWebApp app)
    {
        TestWebResponse response = await app.Get("/authorized/small", request => request.Headers["Authorization"] = $"Bearer {Token("token")}");

        response.Assert.Ok();
        await Answer.Is(Expected.Json("items.small.json"), response);
    }

    // rb:test authorized.denied
    [ModuleTest]
    [Trait("corpus", "authorized.denied")]
    public async Task A_token_one_character_off_is_forbidden(ITestWebApp app)
    {
        TestWebResponse response = await app.Get("/authorized/small", request => request.Headers["Authorization"] = $"Bearer {Token("wrongToken")}");

        response.Assert.Forbidden();
    }

    [ModuleTest]
    public async Task No_token_is_unauthorized(ITestWebApp app)
    {
        (await app.Get("/authorized/small")).Assert.Unauthorized();
    }
}
